types/plugin.ts

```diff
@@ -37,3 +43,302 @@ export const Plugins: Record<PluginID, Plugin> = {
+        in: param.in,
+        description: param.description,
+        required: param.required,
+        schema: param.schema,
+      }));
+
+      const pluginApiOperation: PluginApiOperation = {
+        operationId: operationObject.operationId,
+        nameForModel: plugin.name_for_model,
+        serverUrl: serverUrl,
+        apiPath,
+        method,
+        summary: operationObject.summary,
+        parameters,
+      };
+
+      const contentObject = operationObject.requestBody?.content;
+      if (contentObject) {
+        let content: any = {};
+        for (const key in contentObject) {
+          if (contentObject[key].schema?.hasOwnProperty('$ref')) {
+            contentObject[key].schema = resolveRef(
+              contentObject[key].schema.$ref,
+              document,
+            );
+          }
+          content[key] = {
+            schema: contentObject[key].schema,
+          };
+        }
+        pluginApiOperation.requestBody = {
+          required: operationObject.requestBody.required,
+          content,
+        };
+      }
+
+      pluginApiOperation.responses = {};
+      for (const key in operationObject.responses) {
+        pluginApiOperation.responses[key] = {
+          description: operationObject.responses[key].description,
+        };
+        const response = operationObject.responses[key];
+        const contentObject = response.content;
+        if (contentObject) {
+          let content: any = {};
+          for (const key in contentObject) {
+            if (contentObject[key].schema?.hasOwnProperty('$ref')) {
+              contentObject[key].schema = resolveRef(
+                contentObject[key].schema.$ref,
+                document,
+              );
+            }
+            content[key] = {
+              schema: contentObject[key].schema,
+            };
+          }
+          pluginApiOperation.responses[key].content = content;
+        }
+      }
+
+      pluginApiOperationList[operationObject.operationId] = pluginApiOperation;
+    }
+  }
+
+  return pluginApiOperationList;
+}
+
+export function getOpenAIFunctionFromPluginApiOperation(
+  pluginApiOperation: PluginApiOperation,
+): OpenAIFunction {
+  const {
+    operationId: id,
+    summary,
+    parameters,
+    requestBody,
+  } = pluginApiOperation;
+
+  // Build the properties object for OpenAIFunction
+  const properties: { [key: string]: {} } = {};
+  if (parameters) {
+    for (const parameter of parameters) {
+      properties[parameter.name] = parameter.schema;
+    }
+  }
+
+  // Build the required array for OpenAIFunction
+  const required =
+    parameters
+      ?.filter((parameter) => parameter.required)
+      .map((parameter) => parameter.name) || [];
+
+  if (requestBody) {
+    for (const key in requestBody.content) {
+      for (const propertyKey in requestBody.content[key].schema.properties) {
+        properties[propertyKey] = {
+          type: requestBody.content[key].schema.properties[propertyKey].type,
```

Comment:
This simple mapping doesn't work when plugin schema includes `array` `type` properties. For example, "Doc Maker" (`https://docmaker.level2labs.xyz/.well-known/ai-plugin.json`) plugin.
It's schema looks like sth the following:
```js
{
  schema: {
    type: 'object',
      properties: {
      property_name: {
        type: 'array',
          items: {
          type: 'object',
            properties: {
            heading: { type: 'string' },
            content: { type: 'array', items: { type: 'string' } }
          }
        }
      }
    }
  }
}
```

Simple fix is to do the following, but it might not be optimal solution:
```ts
      for (const propertyKey in requestBody.content[key].schema.properties) {
        const props = requestBody.content[key].schema.properties[propertyKey];

        properties[propertyKey] = {
          description: props.description || '',
          ...props
        };

        if (props.required) {
          required.push(propertyKey);
        }
      }
```
In general, this super long property access (`requestBody.content[key].schema.properties[propertyKey]`) and mapping might need some refactoring.

Reply:
In some cases it seems it's also needed to remove `required` property:
```ts
      for (const propertyKey in requestBody.content[key].schema.properties) {
        const props = requestBody.content[key].schema.properties[propertyKey];

        properties[propertyKey] = {
          description: props.description || '',
          ...props
        };

        delete properties[propertyKey].required;

        if (props.required) {
          required.push(propertyKey);
        }
      }
```